### PR TITLE
CRUD for users, roles, permissions

### PR DESCRIPTION
**Pull request by neerk28**

### CODE REVIEW

File: functions/api/permissions.js
Comment (shrikantdange1208) [inline functions/api/permissions.js]: Could you please define constants for collection and document names. Also, I see you have used permissions. Should we use singular or plural names. I am using singular names.

File: functions/api/roles.js
Comment (shrikantdange1208) [inline functions/api/roles.js]: You can use constants here

File: functions/api/roles.js
Comment (shrikantdange1208) [inline functions/api/roles.js]: .send does not send a response here. You can use sendStatus(204)

File: functions/api/roles.js
Comment (shrikantdange1208) [inline functions/api/roles.js]: Could you please check if the role exists before deleting? In your case it will throw Internal Server Error. Ideally it should be 404.
Same for PUT requests.

File: functions/api/roles.js
Comment (shrikantdange1208) [inline functions/api/roles.js]: For POST, will set create a new entry if the object already exists or it will update it? Created Date will mismatch incase of update.

File: functions/api/roles.js
Comment (shrikantdange1208) [inline functions/api/roles.js]: Do we need validation for Delete? It is similar to Get param

File: functions/package.json
Comment (shrikantdange1208) [inline functions/package.json:22]: Is there any specific reason we are using "firebase-functions": "3.6.1" and not 3.7.2?

**Review by shrikantdange1208 — CHANGES_REQUESTED** [review]
Few common comments I have:
1. Check if entity exists in post, put and delete. (In put and delete it will throw 500)
2. Using constants
3. Not so important. Put requests can use sendStatus(204).

File: functions/api/roles.js
Comment (neerk28) [reply functions/api/roles.js]: Actually I am getting 200, even though that role doesnt exist.

File: functions/api/roles.js
Comment (neerk28) [reply functions/api/roles.js]: will remove and check for doc exists in delete api

File: functions/package.json
Comment (neerk28) [reply functions/package.json:22]: "firebase serve" doesnt work in 3.7.x

File: functions/api/permissions.js
Comment (neerk28) [reply functions/api/permissions.js]: I am using plural form. Naming convention is collection names has to be plural.

File: functions/api/roles.js
Comment (shrikantdange1208) [reply functions/api/roles.js]: Yes, my bad. I also get 200, but it's still not the right status.

File: functions/package.json
Comment (shrikantdange1208) [reply functions/package.json:22]: firebase serve has never worked for me. Does it really works with 3.6.1?

File: functions/api/roles.js
Comment (neerk28) [reply functions/api/roles.js]: modified to check doc exist condition for POST, PUT, DELETE

File: functions/api/roles.js
Comment (neerk28) [reply functions/api/roles.js]: yes, tried mergeFields option- has some issue with new create. so checking doc exist condition.

File: functions/package.json
Comment (neerk28) [reply functions/package.json:22]: yes

**Comment by neerk28** [comment]
addressed review comments. Please review once and approve if it looks good.

File: functions/api/roles.js
Comment (shrikantdange1208) [inline functions/api/roles.js:120]: Small observation: you already have rolesRef which can be reused here.

**Review by shrikantdange1208 — APPROVED** [review]

File: functions/api/roles.js
Comment (neerk28) [reply functions/api/roles.js:120]: oops, may be i missed it. Took care in users.Will correct in my next PR

**Pull request merged**
